Sources/Datadog/Core/Upload/DataUploadStatus.swift

```diff
@@ -0,0 +1,164 @@
+    let userErrorMessage: String?
+
+    /// An optional error logged to the Internal Monitoring feature (if it's enabled).
+    let internalMonitoringError: (message: String, error: Error?, attributes: [String: String]?)?
+}
+
+extension DataUploadStatus {
+    // MARK: - Initialization
+
+    init(httpResponse: HTTPURLResponse, ddRequestID: String?) {
+        let statusCode = HTTPResponseStatusCode(rawValue: httpResponse.statusCode) ?? .unexpected
+
+        self.init(
+            needsRetry: statusCode.needsRetry,
+            userDebugDescription: "[response code: \(httpResponse.statusCode) (\(statusCode)), request ID: \(ddRequestID ?? "(???)")]",
+            userErrorMessage: statusCode == .forbidden ? "⚠️ The client token you provided seems to be invalid." : nil,
+            internalMonitoringError: createInternalMonitoringErrorIfNeeded(for: httpResponse.statusCode, requestID: ddRequestID)
+        )
+    }
+
+    init(networkError: Error) {
+        self.init(
+            needsRetry: true, // retry this upload as it failed due to network transport isse
+            userDebugDescription: "[error: \(DDError(error: networkError).message)]", // e.g. "[error: A data connection is not currently allowed]"
+            userErrorMessage: nil, // nothing actionable for the user
+            internalMonitoringError: createInternalMonitoringErrorIfNeeded(for: networkError)
+        )
+    }
+}
+
+// MARK: - Internal Monitoring
+
+#if DD_SDK_ENABLE_INTERNAL_MONITORING
+/// Looks at the `statusCode` and produces error for Internal Monitoring feature if anything is going wrong.
+private func createInternalMonitoringErrorIfNeeded(
+    for statusCode: Int, requestID: String?
+) -> (message: String, error: Error?, attributes: [String: String]?)? {
+    guard let responseStatusCode = HTTPResponseStatusCode(rawValue: statusCode) else {
+        // If status code is unexpected, do not produce an error for Internal Monitoring - otherwise monitoring may
+        // become too verbose for old installations if we introduce a new status code in the API.
+        return nil
+    }
+
+    switch responseStatusCode {
+    case .accepted, .unauthorized, .forbidden:
+        // These codes mean either success or the user configuration mistake - do not produce error.
+        return nil
+    case .tooManyRequests, .internalServerError, .serviceUnavailable:
+        // These codes mean Datadog service issue - do not produce SDK error as this is already monitored by other means.
+        return nil
+    case .badRequest, .payloadTooLarge, .requestTimeout:
+        // These codes mean that something wrong is happening either in the SDK or on the server - produce an error.
+        return (
+            message: "Data upload finished with status code: \(statusCode)",
+            error: nil,
+            attributes: ["dd_request_id": requestID ?? "(???)"]
+        )
```

Comment:
Here, I'm not sure if `408 Request Timeout` is worth sending error for Internal Monitoring. On one side this could indicate a slow response from Intake (monitored by other means), but on the other the SDK might be lagging on uploads. WDYT?

Reply:
we can merge once this discussion is resolved

Reply:
There's distinction between client (`URLSession`) timeout and server timeout - `429` means the server.

> maybe we can internal-monitor tooManyRequests too.

Good idea, I added `.tooManyRequests` to the list for alerting 👌.